passes/source2il.nim

```diff
@@ -323,36 +334,43 @@ proc fitExpr(c; e: sink Expr, target: SemType): Expr =
   if e.typ == target:
     result = e
   elif e.typ.isSubtypeOf(target):
-    result = Expr(stmts: e.stmts, typ: target)
-    case target.kind
-    of tkUnion:
-      # construct a union
-      let
-        tmp = c.newTemp(target)
-        idx = find(target.elems, e.typ)
-
-      # ignore the type not being found in the union. This indicates that an
-      # error occurred during union construction
-
-      # emit the tag assignment:
-      result.stmts.addStmt Asgn:
-        bu.subTree Field:
-          bu.add Node(kind: Local, val: tmp)
-          bu.add Node(kind: Immediate, val: 0)
-        bu.add Node(kind: IntVal, val: c.literals.pack(idx))
-
-      # emit the value assignment:
-      result.stmts.addStmt Asgn:
-        bu.subTree Field:
+    if e.typ.kind == tkVoid:
+      result = e
+    else:
+      result = Expr(stmts: e.stmts, typ: target)
+      case target.kind
+      of tkUnion:
+        # construct a union
+        let
+          tmp = c.newTemp(target)
+          idx = find(target.elems, e.typ)
+
+        # ignore the type not being found in the union. This indicates that an
+        # error occurred during union construction
+
+        # emit the tag assignment:
+        result.stmts.addStmt Asgn:
           bu.subTree Field:
             bu.add Node(kind: Local, val: tmp)
-            bu.add Node(kind: Immediate, val: 1)
-          bu.add Node(kind: Immediate, val: idx.uint32)
-        genUse(e.expr, bu)
+            bu.add Node(kind: Immediate, val: 0)
+          bu.add Node(kind: IntVal, val: c.literals.pack(idx))
 
-      result.expr = @[Node(kind: Local, val: tmp)]
-    else:
-      unreachable()
+        # emit the value assignment:
+        result.stmts.addStmt Asgn:
+          bu.subTree Field:
+            bu.subTree Field:
+              bu.add Node(kind: Local, val: tmp)
+              bu.add Node(kind: Immediate, val: 1)
+            bu.add Node(kind: Immediate, val: idx.uint32)
+          genUse(e.expr, bu)
+
+        result.expr = @[Node(kind: Local, val: tmp)]
+      of tkUnit:
+        result = Expr(stmts: e.stmts, expr: @[Node(kind: IntVal)], typ: target)
+        result.stmts.addStmt:
+          genDrop(e.expr, e.typ, bu)
```

Comment:
```suggestion
```
This is probably a leftover from debugging? Only `void` fits `unit`, and that's already handled by the `e.typ.kind == tkVoid` branch above.

Reply:
I left it in thinking of tuple sub typing, but that can be added in later.